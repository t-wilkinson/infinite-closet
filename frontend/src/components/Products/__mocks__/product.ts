export const mockProduct = {
  id: 1234567890,
  name: 'Product Name',
  slug: 'product-name',
  designer: { name: 'Designer Name', slug: 'designer-name' },
  shortRentalPrice: 10.0,
  longRentalPrice: 10.0,
  retailPrice: 10.0,
  purchasePrice: 10.0,
  images: [],
  sizes: [{ id: '1', quantity: 1, size: 'MD' }],
  details: 'Product details',
  stylistNotes: 'Stylist notes',
  sizingNotes: 'Sizing notes',
  categories: [],
  colors: [],
  fits: [],
  occasions: [],
  styles: [],
  weather: [],
  categories_: '',
  colors_: '',
  fits_: '',
  occasions_: '',
  styles_: '',
  weather_: '',
  created_by: '',
  updated_at: '',
  published_at: '',
}
