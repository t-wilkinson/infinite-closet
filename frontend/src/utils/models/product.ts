import { StrapiCategory } from './category';
import { StrapiColor } from './color';
import { StrapiDesigner } from './designer';
import { StrapiFile } from './file';
import { StrapiFit } from './fit';
import { StrapiMaterial } from './material';
import { StrapiMetal } from './metal';
import { StrapiOccasion } from './occasion';
import { StrapiSize } from './custom/sizes';
import { StrapiStyle } from './style';
import { StrapiWeather } from './weather';

/**
 * Model definition for Product
 */
export interface StrapiProduct {
  id: string;
  name: string;
  slug: string;
  designer?: StrapiDesigner;
  shortRentalPrice: number;
  longRentalPrice: number;
  retailPrice?: number;
  purchasePrice?: number;
  images: StrapiFile[];
  sizes: StrapiSize[];
  details?: string;
  stylistNotes?: string;
  sizingNotes?: string;
  categories: StrapiCategory[];
  colors: StrapiColor[];
  fits: StrapiFit[];
  occasions: StrapiOccasion[];
  styles: StrapiStyle[];
  weather: StrapiWeather[];
  materials: StrapiMaterial[];
  metals: StrapiMetal[];
  categories_?: string;
  colors_?: string;
  fits_?: string;
  occasions_?: string;
  styles_?: string;
  weather_?: string;
  materials_?: string;
  metals_?: string;
  sizes_?: string;
  created_by: string;
  updated_at: string;
  published_at: string;
}