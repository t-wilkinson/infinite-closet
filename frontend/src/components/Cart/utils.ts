import { createAsyncThunk } from '@reduxjs/toolkit'

import axios from '@/utils/axios'
import { RootState } from '@/utils/store'
import { StrapiOrder } from '@/utils/models'
import {Summary} from '@/types'

import * as helpers from './helpers'
import { Orders, Cart } from './types'

export const getUser = (getState: () => any) =>
  (getState() as RootState).user.data

const toKey = (order: StrapiOrder) => `${order.size}_${order.product.id}`

export default {
  get: createAsyncThunk<Orders, void>('cart/get', async (_, { getState }) => {
    const user = getUser(getState)
    return await helpers.getOrders(user)
  }),
  set: createAsyncThunk<Orders, Orders>(
    'cart/set',
    async (orders, { getState }) => {
      const user = getUser(getState)
      const uniqueOrders: Orders = Object.values(
        orders.reduce((acc, order) => {
          acc[toKey(order)] = order
          return acc
        }, {})
      )
      helpers.setOrders(user, uniqueOrders)
      return orders
    }
  ),
  summary: createAsyncThunk<Summary, void>(
    'cart/summary',
    async (_, { getState }) => {
      const user = getUser(getState)
      const orders = await helpers.getOrders(user)
      let data: Summary
      if (user) {
        data = await axios.post<Summary>(
          `/orders/cart/summary`,
          { orders },
        )
      } else {
        data = await axios.post(`/orders/cart/summary`, { orders }, {withCredentials: false})
      }
      return data
    }
  ),
  count: createAsyncThunk<number, void>(
    'cart/count',
    async (_, { getState }) => {
      const user = getUser(getState)
      if (user) {
        return await axios.get<number>('/orders/cart/count')
      } else {
        const orders = helpers.getGuestOrders()
        return orders.length
      }
    }
  ),
  view: createAsyncThunk<Cart, void>('cart/view', async (_, { getState }) => {
    const user = getUser(getState)
    return helpers.viewOrders(user)
  }),
  history: createAsyncThunk<Cart, void>(
    'cart/history',
    async (_, { getState }) => {
      const user = getUser(getState)
      return helpers.orderHistory(user)
    }
  ),
  update: createAsyncThunk<void, Partial<StrapiOrder>>(
    'cart/update',
    async (order, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.put<void>(`/orders/${order.id}`, order)
      } else {
        await axios.put<void>(`/orders/${order.id}`, order, {withCredentials: false})
        const orders = helpers.getGuestOrders()
        const index = orders.findIndex((v: StrapiOrder) => v.id === order.id)
        orders[index] = { ...orders[index], ...order }
        helpers.setGuestOrders(orders)
      }
    }
  ),
  add: createAsyncThunk<Cart, Partial<StrapiOrder>>(
    'cart/add',
    async (order, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.post<void>('/orders', order)
      } else {
        order = await axios.post<StrapiOrder>('/orders', order, {withCredentials: false})
        const orders = [...helpers.getGuestOrders(), order as StrapiOrder]
        helpers.setGuestOrders(orders)
      }
      return await helpers.viewOrders(user)
    }
  ),
  insert: createAsyncThunk<Cart, Partial<StrapiOrder>[] | Partial<StrapiOrder>>(
    'cart/insert',
    async (inserts, { getState }) => {
      const user = getUser(getState)
      if (!user) {
        return
      }

      if (!Array.isArray(inserts)) {
        inserts = [inserts]
      }

      const orders = await helpers.getUserOrders(user)
      const orderIds = orders
        .concat(inserts as Orders)
        .map((order: StrapiOrder) => order.id)
      helpers.setUserOrders(user, orderIds)
      return await helpers.viewOrders(user)
    }
  ),
  remove: createAsyncThunk<Orders, string>(
    'cart/remove',
    async (id, { getState }) => {
      const user = getUser(getState)
      if (user) {
        await axios.put<void>(`/orders/${id}`, { id, status: 'dropped' })
      } else {
        const orders = helpers.getGuestOrders()
        const filtered = orders.filter((order: StrapiOrder) => order.id !== id)
        helpers.setGuestOrders(filtered)
      }
      return helpers.getOrders(user)
    }
  ),
}
