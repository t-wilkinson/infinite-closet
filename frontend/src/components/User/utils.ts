import { StrapiUser } from '@/types/models'

export const displayUsername = (username: string) => {
  return username ? username.split('-').slice(0, -1).join('-') : ''
}
