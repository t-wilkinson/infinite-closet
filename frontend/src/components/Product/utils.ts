import { getURL } from '@/utils/axios'
import { StrapiProduct } from '@/types/models'

export const productImageProps = (product: StrapiProduct) => {
  const mainImage = product.images[0]
  return {
    src: getURL(mainImage.formats.thumbnail?.url || mainImage.url),
    alt:
      mainImage.alternativeText ||
      product.designer ? `${product.name} by ${product.designer.name}` : product.name,
  }
}

export const productRentalPrice = (product: StrapiProduct, rentalLength: 'short' | 'long') => {
  switch (rentalLength) {
    case 'short':
      return product.shortRentalPrice
    case 'long':
      return product.longRentalPrice
    default:
      return undefined
  }
}
