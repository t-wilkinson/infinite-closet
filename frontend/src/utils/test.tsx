import React from 'react'
import { configureStore } from '@reduxjs/toolkit'
import { storeOptions } from '@/utils/store'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'

function render(
  ui: React.ReactElement<any>,
  {
    initialState = {},
    store = configureStore({ ...storeOptions, preloadedState: initialState }),
    ...renderOptions
  } = {},
) {
  store.dispatch = jest.fn()

  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  let component = rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
  component.store = store
  return component
}

export * from '@testing-library/react'
export { render }
