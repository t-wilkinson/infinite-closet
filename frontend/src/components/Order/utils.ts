import { toast } from 'react-toastify'

import axios from '@/utils/axios'
import * as sizing from '@/utils/sizing'
import { currency } from '@/utils/config'
import { productRentalPrice } from '@/Product/utils'

import { OrderUtils } from './slice'

const prepareOrder = ({ product, fields, user }) => {
  const size = sizing.get(product.sizes, fields.get('size').value)
  const selectedDate = fields.get('selectedDate').value
  const rentalLength = fields.get('rentalLength').value

  const order: any = {
    user: user ? user.id : null,
    size: sizing.unnormalize(size.size),
    product: product.id,
    expectedStart: selectedDate ? selectedDate.toJSON() : undefined,
    rentalLength: rentalLength ? rentalLength : undefined,
  }
  return order
}

const productToItem = (product) => {
  return {
    item_id: product.id,
    item_name: product.name,
    item_brand: product.designer.name,
    // item_category: order.,
    price: product.retailPrice,
    quantity: 1,
  }
}

export const addToCart = async ({
  product,
  fields,
  user,
  dispatch,
  analytics,
}) => {
  const order = { ...prepareOrder({ product, fields, user }), status: 'cart' }

  return dispatch(OrderUtils.add(order))
    .then(() => {
      toast.success(`Successfully added to cart.`, {
        autoClose: 1500,
        hideProgressBar: true,
      }),
        analytics.logEvent('add_to_cart', {
          currency,
          value: productRentalPrice(product, order.rentalLength),
          user: user ? user.email : 'guest',
          items: [
            productToItem(product)
          ],
        })
    })
    .catch(() => {
      toast.error(`Unable to add to cart`, {})
    })
}

export const addToFavorites = async ({
  product,
  fields,
  user,
  dispatch,
  analytics,
}) => {
  try {
    const order = { ...prepareOrder({ product, fields, user }), status: 'list' }
    await dispatch(OrderUtils.add(order))
    await dispatch(OrderUtils.favorites())
    toast.success(`Successfully added to favorites.`, {
      autoClose: 1500,
      hideProgressBar: true,
    })
    analytics.logEvent('add_to_wishlist', {
      currency,
      value: productRentalPrice(product, order.rentalLength),
      items: [
         productToItem(product)
      ],
    })
  } catch (e) {
    console.log(e)
    toast.error(
      `Ran into an issue adding to favorites. We'll have this fixed soon!`,
      {}
    )
  }
}

export const removeOrderItem = async ({ dispatch, analytics, order }) => {
  const { product } = order
  await dispatch(OrderUtils.remove(order.id))
  await dispatch(OrderUtils.view())
  analytics.logEvent('remove_from_cart', {
    currency,
    value: productRentalPrice(product, order.rentalLength),
    items: [
      productToItem(product)
    ],
  })
}

export const toggleInsurance = ({ dispatch, order }) => {
  dispatch(
    OrderUtils.update({ id: order.id, insurance: !order.insurance })
  ).then(() => dispatch(OrderUtils.view()))
}

export const moveToFavorites = ({ order, dispatch }) => {
  dispatch(OrderUtils.update({ id: order.id, status: 'list' })).then(() => {
    dispatch(OrderUtils.view())
    dispatch(OrderUtils.favorites())
  })
}

export const getValidOrderDates = ({ product, dates, size, rentalLength }) =>
  axios
    .post(
      '/orders/dates/valid',
      {
        dates,
        product: product.id,
        size: (sizing.get(product.sizes, size) || product.sizes[0]).size,
        rentalLength,
      },
      { withCredentials: false }
    )
    .then((data) => data.valid)
    .catch((err) => {
      toast.error(`We ran into an issue verifying the available dates.`)
    })

export const availableOrderDates = async ({
  date,
  product,
  size,
  rentalLength,
}) => {
  const curDay = date.hour(12).date(1)
  const dates = [
    curDay.date(0).subtract(1, 'month'),
    curDay.date(0),
    curDay.add(1, 'month').date(0),
    curDay.add(2, 'month').date(0),
    curDay.add(3, 'month').date(0),
  ].reduce((acc, day) => {
    Array(day.date())
      .fill(0)
      .map((_, i) => {
        acc.push(day.set('date', i + 1))
      })
    return acc
  }, [])

  // TODO: should not refetch everytime new date overed over
  return await getValidOrderDates({ product, size, rentalLength, dates })
}

const rentalLengths = {
  short: 4 - 1,
  long: 8 - 1,
}

export const getRentalLength = (rentalLength: 'short' | 'long') =>
  rentalLengths[rentalLength]
