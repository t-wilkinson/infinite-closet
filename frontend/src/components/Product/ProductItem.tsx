import React from 'react'
import Link from 'next/link'
import Image from 'next/image'

import { getURL } from '@/utils/axios'
import useAnalytics from '@/utils/useAnalytics'
import { useDispatch, useSelector } from '@/utils/store'
import * as sizing from '@/utils/sizing'
import { OrderUtils, addToFavorites } from '@/Order'
import { StrapiProduct } from '@/types/models'
import { Icon, iconHeartFill } from '@/Components/Icons'
import { useFields } from '@/Form'

import { AddToCartFields } from './ProductPage/types'
import styles from './Products.module.css'

export const ProductWrapper = ({ children = null }) => (
  <div className="md:px-4 md:py-6 w-1/2 lg:w-1/3 ">
    <div
      className="w-full
        border-transparent border hover:border-gray  "
    >
      <div className="w-full h-full p-2">{children}</div>
    </div>
  </div>
)

const AddToFavorites = ({product}) => {
  const [addedFavorite, setAddedFavorite] = React.useState(false)

  const favorites = useSelector((state) => state.orders.favorites)
  const isFavorite = favorites.find(o => o.product.id === product.id)
  const dispatch = useDispatch()
    const analytics = useAnalytics()
  const user = useSelector((state) => state.user.data)

  const fields = useFields<AddToCartFields>({
    visible: { },
    size: { constraints: 'required', default: product.sizes[0]?.size },
    selectedDate: {
      label: 'Rental Date',
      constraints: 'required',
      default: null,
    },
    rentalLength: { constraints: 'required', default: 'short' },
    rentType: { default: 'OneTime' },
  })

  React.useEffect(() => {
    dispatch(OrderUtils.favorites())
  }, [user])

  if (!product.sizes[0]?.size) {
    return null
  }

  return <button
    className="stroke-pri p-2 border border-transparent hover:border-gray-light"
      disabled={addedFavorite}
      onClick={() => {
        addToFavorites({fields, dispatch, analytics, user, product})
        .then(() => setAddedFavorite(true))
      }}
  >
    <Icon className={`${isFavorite ? 'text-red' : 'text-transparent'} stroke-2 stroke-pri`}
    icon={iconHeartFill} size={32} />
  </button>
}

export const WrappedProductItem = ({ product }: { product: StrapiProduct }) =>
  <ProductWrapper>
    <ProductItem product={product} />
  </ProductWrapper>


export const ProductItem = ({ product }: { product: StrapiProduct }) => {
  const analytics = useAnalytics()

  return (
    <>
      <div
        className="relative w-full"
        style={{
          paddingTop: '135%',
        }}
      >
        <Link href={`/shop/${product.customDesignerName || product.designer?.slug || 'unknown'}/${product.slug}`}>
          <a
            className="absolute inset-0"
            onClick={() =>
              analytics.logEvent('select_item', {
                type: 'products.select-item',
              })
            }
          >
            <ProductImages product={product} />
          </a>
        </Link>
        <div className="absolute top-0 right-0">
          <AddToFavorites product={product} />
        </div>
      </div>
      <ProductInfo product={product} />
      </>
  )
}

const ProductImageWrapper = ({ image, hover, show = false, hide = false }) => (
  <div className="absolute inset-0 z-0" key={image.url}>
    <div
      className={`relative w-full h-full
        ${
          !hover && hide
            ? ''
            : !hover
            ? 'hidden'
            : show
            ? styles['animate-show'] // only show first image if not hovering
            : hide
            ? styles['animate-hide']
            : 'hidden'
        }
    `}
    >
      <ProductImage
        alt={image.alternativeText}
        src={getURL(image.url)}
        ratio={image.height / image.width}
      />
    </div>
  </div>
)

export const ProductImages = ({ product }) => {
  const [hoverInterval, setHoverInterval] = React.useState<number>()
  const [index, setIndex] = React.useState<number>(0)
  const wrap = (i: number) => i % product.images.length
  const rotate = () => {
    setIndex((i) => wrap(i + 1))
  }

  const getSrc = (index: number) => {
    const image = product.images[index]
    return { ...image, ...(image.formats?.medium || null) }
  }

  return (
    <div
      className="w-full h-full relative"
      onMouseOver={() => {
        if (!hoverInterval) {
          setHoverInterval(window.setInterval(rotate, 2000))
        }
      }}
      onMouseLeave={() => {
        setIndex(0)
        window.clearInterval(hoverInterval)
        setHoverInterval(undefined)
      }}
    >
      {product.images.map((_: any, i: number) => {
        // TODO: What about prodcuts with only one image
        const image = getSrc(i)
        return <ProductImageWrapper
          key={i}
          hover={hoverInterval}
          hide={i === index}
          show={i === wrap(index + 1)}
          image={image}
        />
      })}
    </div>
  )
}

// TODO: use nextjs caching
const myLoader = ({ src, width, quality }) => {
  return `${src}?w=${width}&q=${quality || 75}`
  // return `https://infinitecloset.treywilkinson.com/_next/image?url=${encodeURIComponent(
  //   src,
  // )}&w=${width}&q=${quality || 75}`
}

const ProductImage = ({ alt, src, ratio }) => (
  <div className="relative h-full">
    <Image
      loader={myLoader}
      alt={alt}
      src={src}
      layout="fill"
      objectFit={ratio > 1.4 ? 'contain' : 'cover'}
    />
  </div>
)

const rentalPrice = (low: number, high: number): string => {
  if (!low && !high) {
    return '£-'
  }
  return low === high ? `£${low}` : `£${low} - £${high}`
}

export const ProductInfo = ({ product }) => (
  <div className="w-full flex-row justify-between mt-4 relative text-sm leading-snug">
    <div className="flex-grow">
      <div className="inline-block">
        <Link href={`/designers/${product.customDesignerName || product.designer?.slug}`}>
          <a>
            <span className="font-bold hover:underline">
              {product.designer?.name}
            </span>
          </a>
        </Link>
      </div>

      <div className="inline-block">
        <Link href={`/shop/${product.customDesignerName || product.designer?.slug}/${product.slug}`}>
          <a className="hover:underline">
            <span>{product.name}</span>
          </a>
        </Link>
      </div>

      {product.user ?
        <span className="mt-6 font-bold">
          Your outfit
        </span>
        :
      <>
      <span className="text-sm">{sizing.range(product.sizes).join(', ')}</span>
      <div className="h-2" />

      <div className="flex-col md:flex-row">
        <span className="font-bold">
          {rentalPrice(product.shortRentalPrice, product.longRentalPrice)}
        </span>
        <span>
          <span className="hidden md:inline">&nbsp;{'| '}</span>
          <span className="text-gray line-through">
            £{product.retailPrice || '-'} retail
          </span>
        </span>
      </div>
      </>
      }
    </div>
  </div>
)

export default ProductItem
