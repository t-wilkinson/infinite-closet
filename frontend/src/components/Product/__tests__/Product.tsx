import React from 'react'
import * as t from '@/utils/test'
import ProductItem from '@/Product/ProductItem'
import { product } from '../__mocks__/product.ts'
import renderer from 'react-test-renderer';

describe('ProductItem', () => {
  test('.', () => {
    const tree = renderer.create(t.render(<ProductItem product={product} />))
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
