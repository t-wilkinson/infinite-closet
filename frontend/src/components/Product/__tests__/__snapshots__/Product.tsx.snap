// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem . 1`] = `
<ProductItem
  product={
    Object {
      "categories": Array [],
      "categories_": "dresses,midi,clothing",
      "colors": Array [],
      "colors_": "blue",
      "created_at": "2021-05-26T23:01:06.486Z",
      "created_by": null,
      "designer": Object {
        "created_at": "2021-05-26T23:01:06.488Z",
        "created_by": null,
        "description": null,
        "id": "4",
        "name": "GAÃ‚LA",
        "notes": null,
        "products": undefined,
        "published_at": "2021-05-27T14:48:45.089Z",
        "slug": "gaala",
        "updated_at": "2021-05-27T14:48:45.271Z",
        "updated_by": null,
      },
      "details": "This dreamy mid dress fluidly lauds every curve in its simple elegance. ",
      "fits": Array [],
      "fits_": "",
      "id": "13",
      "images": Array [
        Object {
          "alternativeText": "",
          "caption": "",
          "created_at": "2021-05-12T17:27:43.035Z",
          "created_by": 1,
          "ext": ".jpg",
          "formats": Object {
            "large": Object {
              "ext": ".jpg",
              "hash": "large_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933",
              "height": 1000,
              "mime": "image/jpeg",
              "name": "large_287C2800-6900-4749-B194-1AE2E535ACDC_1512x.jpg",
              "path": null,
              "size": 89.26,
              "url": "/uploads/large_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933.jpg",
              "width": 667,
            },
            "medium": Object {
              "ext": ".jpg",
              "hash": "medium_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933",
              "height": 750,
              "mime": "image/jpeg",
              "name": "medium_287C2800-6900-4749-B194-1AE2E535ACDC_1512x.jpg",
              "path": null,
              "size": 57.62,
              "url": "/uploads/medium_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933.jpg",
              "width": 500,
            },
            "small": Object {
              "ext": ".jpg",
              "hash": "small_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933",
              "height": 500,
              "mime": "image/jpeg",
              "name": "small_287C2800-6900-4749-B194-1AE2E535ACDC_1512x.jpg",
              "path": null,
              "size": 29.87,
              "url": "/uploads/small_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933.jpg",
              "width": 333,
            },
            "thumbnail": Object {
              "ext": ".jpg",
              "hash": "thumbnail_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933",
              "height": 156,
              "mime": "image/jpeg",
              "name": "thumbnail_287C2800-6900-4749-B194-1AE2E535ACDC_1512x.jpg",
              "path": null,
              "size": 4.74,
              "url": "/uploads/thumbnail_287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933.jpg",
              "width": 104,
            },
          },
          "hash": "287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933",
          "height": 2268,
          "id": 50,
          "mime": "image/jpeg",
          "name": "287C2800-6900-4749-B194-1AE2E535ACDC_1512x.jpg",
          "previewUrl": null,
          "provider": "local",
          "provider_metadata": null,
          "size": 223.55,
          "updated_at": "2021-05-12T17:27:43.079Z",
          "updated_by": 1,
          "url": "/uploads/287_C2800_6900_4749_B194_1_AE_2_E535_ACDC_1512x_a38080e933.jpg",
          "width": 1512,
        },
        Object {
          "alternativeText": "",
          "caption": "",
          "created_at": "2021-05-12T17:27:43.355Z",
          "created_by": 1,
          "ext": ".jpg",
          "formats": Object {
            "large": Object {
              "ext": ".jpg",
              "hash": "large_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec",
              "height": 1000,
              "mime": "image/jpeg",
              "name": "large_5974D912-CD5B-42C6-B4D3-5C2A572ADF7B_3000x.jpg",
              "path": null,
              "size": 88.83,
              "url": "/uploads/large_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec.jpg",
              "width": 667,
            },
            "medium": Object {
              "ext": ".jpg",
              "hash": "medium_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec",
              "height": 750,
              "mime": "image/jpeg",
              "name": "medium_5974D912-CD5B-42C6-B4D3-5C2A572ADF7B_3000x.jpg",
              "path": null,
              "size": 56.87,
              "url": "/uploads/medium_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec.jpg",
              "width": 500,
            },
            "small": Object {
              "ext": ".jpg",
              "hash": "small_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec",
              "height": 500,
              "mime": "image/jpeg",
              "name": "small_5974D912-CD5B-42C6-B4D3-5C2A572ADF7B_3000x.jpg",
              "path": null,
              "size": 30.37,
              "url": "/uploads/small_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec.jpg",
              "width": 333,
            },
            "thumbnail": Object {
              "ext": ".jpg",
              "hash": "thumbnail_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec",
              "height": 156,
              "mime": "image/jpeg",
              "name": "thumbnail_5974D912-CD5B-42C6-B4D3-5C2A572ADF7B_3000x.jpg",
              "path": null,
              "size": 4.75,
              "url": "/uploads/thumbnail_5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec.jpg",
              "width": 104,
            },
          },
          "hash": "5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec",
          "height": 3000,
          "id": 52,
          "mime": "image/jpeg",
          "name": "5974D912-CD5B-42C6-B4D3-5C2A572ADF7B_3000x.jpg",
          "previewUrl": null,
          "provider": "local",
          "provider_metadata": null,
          "size": 282.38,
          "updated_at": "2021-05-12T17:27:43.364Z",
          "updated_by": 1,
          "url": "/uploads/5974_D912_CD_5_B_42_C6_B4_D3_5_C2_A572_ADF_7_B_3000x_7e06dafaec.jpg",
          "width": 2000,
        },
        Object {
          "alternativeText": "",
          "caption": "",
          "created_at": "2021-05-12T17:27:43.145Z",
          "created_by": 1,
          "ext": ".jpg",
          "formats": Object {
            "large": Object {
              "ext": ".jpg",
              "hash": "large_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c",
              "height": 1000,
              "mime": "image/jpeg",
              "name": "large_F4C25584-E63A-4EED-85C5-0F3B6DA0D2BB_1512x.jpg",
              "path": null,
              "size": 63,
              "url": "/uploads/large_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c.jpg",
              "width": 667,
            },
            "medium": Object {
              "ext": ".jpg",
              "hash": "medium_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c",
              "height": 750,
              "mime": "image/jpeg",
              "name": "medium_F4C25584-E63A-4EED-85C5-0F3B6DA0D2BB_1512x.jpg",
              "path": null,
              "size": 41.13,
              "url": "/uploads/medium_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c.jpg",
              "width": 500,
            },
            "small": Object {
              "ext": ".jpg",
              "hash": "small_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c",
              "height": 500,
              "mime": "image/jpeg",
              "name": "small_F4C25584-E63A-4EED-85C5-0F3B6DA0D2BB_1512x.jpg",
              "path": null,
              "size": 22.23,
              "url": "/uploads/small_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c.jpg",
              "width": 333,
            },
            "thumbnail": Object {
              "ext": ".jpg",
              "hash": "thumbnail_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c",
              "height": 156,
              "mime": "image/jpeg",
              "name": "thumbnail_F4C25584-E63A-4EED-85C5-0F3B6DA0D2BB_1512x.jpg",
              "path": null,
              "size": 4.07,
              "url": "/uploads/thumbnail_F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c.jpg",
              "width": 104,
            },
          },
          "hash": "F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c",
          "height": 2268,
          "id": 51,
          "mime": "image/jpeg",
          "name": "F4C25584-E63A-4EED-85C5-0F3B6DA0D2BB_1512x.jpg",
          "previewUrl": null,
          "provider": "local",
          "provider_metadata": null,
          "size": 144.48,
          "updated_at": "2021-05-12T17:27:43.209Z",
          "updated_by": 1,
          "url": "/uploads/F4_C25584_E63_A_4_EED_85_C5_0_F3_B6_DA_0_D2_BB_1512x_faa752580c.jpg",
          "width": 1512,
        },
      ],
      "longRentalPrice": 62,
      "name": "Amandine",
      "occasions": Array [],
      "occasions_": "cocktail,date-night,brunch",
      "published_at": "2021-07-27T08:58:10.946Z",
      "purchasePrice": 230,
      "retailPrice": 230,
      "shortRentalPrice": 37,
      "sizes": Array [],
      "sizes_": "XS,S,L,XL,XXS,M",
      "sizingNotes": null,
      "slug": "amandine",
      "styles": Array [],
      "styles_": "tea-length,sleeveless,sweetheart",
      "stylistNotes": "97% silk 3% elastane  and Made in Belarus from upcycled material",
      "updated_at": "2021-07-27T08:58:14.383Z",
      "updated_by": 1,
      "weather": Array [],
      "weather_": "warm-weather",
    }
  }
/>
`;
